rfc/src/rfcs/0014-harness-partition.md

```diff
@@ -0,0 +1,204 @@
+
+```
+Checking partition #1 (*input > 0) of harness partitioned_proof...
+CBMC 6.7.1 (cbmc-6.7.1)
+/* MORE CBMC OUTPUT */
+RESULTS:
+Check 1: partitioned_proof_partition_1.assertion.1
+     - Status: SUCCESS
+     - Description: "assertion failed: target_fn(input) > 0"
+     - Location: src/pre_partition.rs:8:5 in function partitioned_proof_partition_1
+
+Checking partition #2 (*input < 0) of harness partitioned_proof...
+CBMC 6.7.1 (cbmc-6.7.1)
+/* MORE CBMC OUTPUT */
+RESULTS:
+Check 1: partitioned_proof_partition_2.assertion.1
+     - Status: SUCCESS
+     - Description: "assertion failed: target_fn(input) > 0"
+     - Location: src/pre_partition.rs:8:5 in function partitioned_proof_partition_2
+
+Checking coverage of partitions for harness partitioned_proof...
+CBMC 6.7.1 (cbmc-6.7.1)
+/* MORE CBMC OUTPUT */
+RESULTS:
+Check 1: partitioned_proof_coverage.assertion.1
+     - Status: FAILURE
+     - Description: "partitions for partitioned_proof do not cover all possible values of the i32 type."
+     - Location: src/pre_partition.rs:11:5 in function partitioned_proof_coverage
+```
+
+Note that the coverage check fails, since we're missing the `i = 0` case.
+
+## Software Design
+
+This change would introduce the new `kani::partition` function, with the following signature:
+
+```rust
+pub fn partition<T: Arbitrary, R, const N: usize>(
+    conditions: [fn(&T) -> bool; N],
+    and_run: fn(T) -> R, // the "partition closure"
+) -> R
+```
+
+This signature makes a few key design decisions:
+- *It takes in an array of `conditions` with the specific length `const N`*, ensuring the number of harnesses we have to generate is knowable at compile time when we're doing partition generation.
+- *Partition conditions & the partition closure are represented by function pointers*.
+This allows the user to provide closures of the same signature, but only if those closures do not capture any state.
+Allowing conditions closures to capture runtime state like local variables would be unwise as partitions are conceptually static divisions of a proof which should not depend on runtime values.
+- *The function returns the same type `R` as the underlying partition closure*, with the thought being that value could then be used in later computation without affecting the mechanics of a partition.
+- *The partition variable is bound by `T: Arbitrary`*.
+This would allow us to construct a new non-deterministic value with `kani::any()` that can then be passed to the partition closure.
+
+When the `kani-compiler` encounters a call to this function, it will transparently replace the entire `#[kani::proof()]` that called it with a set of new generated functions that, as a whole, will represent the partitioned proof. This new partitioned proof would contain:
```

Comment:
```suggestion
When the `kani-compiler` encounters a call to this function, it will transparently replace the entire `#[kani::proof]` that called it with a set of new generated functions that, as a whole, will represent the partitioned proof. This new partitioned proof would contain:
```